### ISSUE TITLE
Data streaming 

### ISSUE THREAD

**Issue by pdonorio** [EUDAT-B2STAGE/http-api#71]
We may now avoid the caching of elements in upload and download.

[Here](https://blog.pelicandd.com/article/80/streaming-input-and-output-in-flask) is an example of Streaming input and output in Flask.

@mdantonio has already implemented something similar:

```
from flask import stream_with_context, Response

f = open(path, "rb")
return Response(
    stream_with_context(read_in_chunks(f)),
    mimetype=yourMimeType)


    def read_in_chunks(file_object, chunk_size=1024):
        """
        Lazy function (generator) to read a file piece by piece.
        Default chunk size: 1k.
        """
        while True:
            data = file_object.read(chunk_size)
            if not data:
                break
            yield data
```

**Comment by muccix** [comment]
Some useful posts:
https://github.com/pallets/flask/issues/2086#issuecomment-261962321
https://stackoverflow.com/questions/20015550/read-file-data-without-saving-it-in-flask/20017830

The first approach could be better in performance but requires to manage both form and bonary-data upload

I will try the second approach and then evaluate the performance

**Comment by muccix** [comment]
Some results uploading a 200MB file:

- Upload with `chunk = myfile.read(1024)` and writing on irods:  1m13.107s
- Upload with `chunk = myfile.read(1024)` WITHOUT writing on irods:  1m19.107s (!)
- Upload with `chunk = myfile.read(4096)` and writing on irods:  1m17.107s
- Upload with `chunk = myfile.read(4096)` eWITHOUT writing on irods:  1m14.107s (!)

**Comment by mdantonio** [comment]
I created a Downloader class into the backend repository and pushed the functions for both streamed and partial download that I implemented for the IMC project

https://github.com/rapydo/http-api/blob/master/rapydo/services/download.py

Please note that I also deprecated the download function in the Uploder class:

https://github.com/rapydo/http-api/blob/master/rapydo/services/uploader.py#L56-L72

This function is now available in the new Downloader class


```
from rapydo.services.download import Downloader
[...]
def get():
    [...]
    download = Downloader()
    return download.send_file_partial(file_path, mime)
```

Let me know if you need some modification (for example to handle the mime type directly inside the function: we could add a None default for mime and in this case try automatically retrieve it)

**Comment by muccix** [comment]
See 159d9a5f91a613b8b44b664680fb66ed7573b1f7 and https://github.com/rapydo/http-api/commit/60bb011d0eff8e0ed3d7a24a4ab97cb4d5d328ab
Now both Form and Streaming upload are supported.

Streaming is much faster (1GB, 11 sec with Form, 2 sec with Streaming).
Moreover, the file is not loaded in memory on the client and not saved on a temporary file on the HTTP-API server.

Transferred 10GB in 30 seconds, both with CURL and python requests.

Example2:
```bash
$ curl -T file \
  -H "Authorization: Bearer <auth_token>"
  -H "Content-Type: application/octet-stream"
  <http_server:port>/api/registered/path/to/directory/filename
```

```python
# Streaming upload 'myfile' in '/path/to/directory/filename'
import requests

headers = {"Authorization":"Bearer <token>", "Content-Type":"application/octet-stream"}

with open('/tmp/file', 'rb') as f:
    requests.put('<http_server:port>/api/registered/path/to/directory/filename', data=f, headers=headers)
```

**Comment by muccix** [comment]
Maybe could be good to make configurable the chunk_size?

**Comment by pdonorio** [comment]
> Maybe could be good to make configurable the chunk_size?

Yes, we should get that value from the configuration.
I can help you with that :)

**Comment by muccix** [comment]
Added streaming download from iRODS (no file stored on the HTTP-API) server.

**Comment by pdonorio** [comment]
> we should get that value from the configuration

[There](https://github.com/rapydo/http-api/blob/0.5.1/restapi/flask_ext/flask_irods/client.py#L26) we have it. This will be available in the upcoming 0.5.2 release this week.

Should we close this issue?
